morphia/src/main/java/org/mongodb/morphia/IndexHelper.java

```diff
@@ -0,0 +1,492 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.mongodb.morphia;
+
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.MongoDatabase;
+import org.bson.BsonDocument;
+import org.bson.BsonDocumentWriter;
+import org.bson.codecs.Encoder;
+import org.bson.codecs.EncoderContext;
+import org.mongodb.morphia.annotations.Collation;
+import org.mongodb.morphia.annotations.Field;
+import org.mongodb.morphia.annotations.Index;
+import org.mongodb.morphia.annotations.IndexOptions;
+import org.mongodb.morphia.annotations.Indexed;
+import org.mongodb.morphia.annotations.Indexes;
+import org.mongodb.morphia.annotations.NotSaved;
+import org.mongodb.morphia.annotations.Reference;
+import org.mongodb.morphia.annotations.Serialized;
+import org.mongodb.morphia.annotations.Text;
+import org.mongodb.morphia.logging.Logger;
+import org.mongodb.morphia.logging.MorphiaLoggerFactory;
+import org.mongodb.morphia.mapping.MappedClass;
+import org.mongodb.morphia.mapping.MappedField;
+import org.mongodb.morphia.mapping.Mapper;
+import org.mongodb.morphia.mapping.MappingException;
+import org.mongodb.morphia.utils.IndexType;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.TimeUnit;
+
+import static java.lang.String.format;
+import static java.lang.reflect.Proxy.newProxyInstance;
+import static java.util.Arrays.asList;
+import static java.util.Collections.emptyList;
+import static java.util.Collections.singletonList;
+import static org.mongodb.morphia.utils.IndexType.fromValue;
+
+class IndexHelper {
+    private static final Logger LOG = MorphiaLoggerFactory.get(IndexHelper.class);
+    private static final EncoderContext ENCODER_CONTEXT = EncoderContext.builder().build();
+
+    private final AdvancedDatastore datastore;
+    private final Mapper mapper;
+    private final MongoDatabase database;
+
+    IndexHelper(final AdvancedDatastore datastore, final Mapper mapper, final MongoDatabase database) {
+        this.datastore = datastore;
+        this.mapper = mapper;
+        this.database = database;
+    }
+
+    @SuppressWarnings("deprecation")
+    protected com.mongodb.client.model.IndexOptions convert(final IndexOptions options, final boolean background) {
+        if (options.dropDups()) {
```

Comment:
This will cause runtime errors for existing users, right?  What's the justification for that without a deprecation period?  I know the server dropped support for this, but if they didn't upgrade their server...